File: 00_Inbox/自動化スクリプト/create_weekly_note.js
#!/usr/bin/env node
/**
 * ウィークリーノート自動生成スクリプト
 * 毎週土曜日に実行し、その週（月〜日）のデイリーノートを集約したウィークリーノートを作成
 * 
 * 使い方:
 *   node create_weekly_note.js
 *   node create_weekly_note.js 2026-W03  # 特定の週を指定
 */

const fs = require('fs');
const path = require('path');

// ===== 設定 =====
const VAULT_PATH = '/Users/sota.yamanaka/Documents/Obsidian Vault';
const DAILY_NOTE_FOLDER = '01_Fleeting Note/日記';
const WEEKLY_NOTE_FOLDER = '01_Fleeting Note/日記/週次';

// ===== ユーティリティ関数 =====

/**
 * ISO週番号を取得
 */
function getISOWeek(date) {
    const d = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
    const dayNum = d.getUTCDay() || 7;
    d.setUTCDate(d.getUTCDate() + 4 - dayNum);
    const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
    return Math.ceil((((d - yearStart) / 86400000) + 1) / 7);
}

/**
 * ISO週の開始日（月曜日）を取得
 */
function getWeekStart(year, week) {
    const jan4 = new Date(year, 0, 4);
    const dayOfWeek = jan4.getDay() || 7;
    const weekStart = new Date(jan4);
    weekStart.setDate(jan4.getDate() - dayOfWeek + 1 + (week - 1) * 7);
    return weekStart;
}

/**
 * 日付をYYYY-MM-DD形式でフォーマット
 */
function formatDate(date) {
    const y = date.getFullYear();
    const m = String(date.getMonth() + 1).padStart(2, '0');
    const d = String(date.getDate()).padStart(2, '0');
    return `${y}-${m}-${d}`;
}

/**
 * 曜日を取得
 */
function getDayOfWeek(date) {
    const days = ['日', '月', '火', '水', '木', '金', '土'];
    return days[date.getDay()];
}

/**
 * デイリーノートの内容をパース
 */
function parseDailyNote(content) {
    // YAMLフロントマターを除去
    const bodyContent = content.replace(/^---[\s\S]*?---\s*/, '');
    
    // ##見出しで分割
    const sections = {};
    const sectionOrder = [];
    const parts = bodyContent.split(/^##\s+/m);
    
    parts.forEach(part => {
        if (!part.trim()) return;
        
        const lines = part.split('\n');
        const headerName = lines[0].trim();
        const body = lines.slice(1).join('\n').trim();
        
        if (!body) return;
        
        // 内容が空白行のみの場合はスキップ
        if (body.replace(/^\s*-\s*$/gm, '').trim() === '') return;
        
        if (!sections[headerName]) {
            sections[headerName] = body;
            sectionOrder.push(headerName);
        }
    });
    
    return { sections, sectionOrder };
}

/**
 * ウィークリーノートを生成
 */
function generateWeeklyNote(year, week) {
    const weekStart = getWeekStart(year, week);
    const weekEnd = new Date(weekStart);
    weekEnd.setDate(weekStart.getDate() + 6);
    
    const dailyNotePath = path.join(VAULT_PATH, DAILY_NOTE_FOLDER);
    
    // セクションごとの集約データ
    const accumulatedSections = {};
    const allHeadersOrder = [];
    let foundNotes = 0;
    
    // 7日間のデイリーノートを処理
    for (let i = 0; i < 7; i++) {
        const currentDate = new Date(weekStart);
        currentDate.setDate(weekStart.getDate() + i);
        const dateStr = formatDate(currentDate);
        const dayOfWeek = getDayOfWeek(currentDate);
        const filePath = path.join(dailyNotePath, `${dateStr}.md`);
        
        if (fs.existsSync(filePath)) {
            foundNotes++;
            const content = fs.readFileSync(filePath, 'utf8');
            const { sections, sectionOrder } = parseDailyNote(content);
            
            sectionOrder.forEach(headerName => {
                if (!accumulatedSections[headerName]) {
                    accumulatedSections[headerName] = [];
                    allHeadersOrder.push(headerName);
                }
                
                const body = sections[headerName];
                // 引用形式で日付付きで追加
                const quotedBody = body.split('\n').map(l => `> ${l}`).join('\n');
                accumulatedSections[headerName].push(
                    `> [!quote] [[${dateStr}]] (${dayOfWeek})\n${quotedBody}`
                );
            });
        }
    }
    
    // ノートの生成
    let output = `---
tags: [weekly-review]
week: ${year}-W${String(week).padStart(2, '0')}
date-range: ${formatDate(weekStart)} ~ ${formatDate(weekEnd)}
created: ${formatDate(new Date())}
---

# 📆 Weekly Note: ${year}-W${String(week).padStart(2, '0')}
**期間:** ${formatDate(weekStart)} 〜 ${formatDate(weekEnd)}
**デイリーノート数:** ${foundNotes}/7

---

`;
    
    if (allHeadersOrder.length === 0) {
        output += '> [!warning] 今週のデイリーノートが見つかりませんでした、または内容が空です。\n';
    } else {
        allHeadersOrder.forEach(header => {
            output += `## ${header}\n\n`;
            output += accumulatedSections[header].join('\n\n');
            output += '\n\n';
        });
    }
    
    // 週の振り返りセクション
    output += `---

## 📊 週の振り返り

### 今週のハイライト
- 

### 来週に向けて
- 

`;
    
    return {
        content: output,
        fileName: `${year}-W${String(week).padStart(2, '0')}.md`,
        foundNotes
    };
}

// ===== メイン処理 =====
function main() {
    // コマンドライン引数から週を取得、または現在の週を使用
    let year, week;
    
    const arg = process.argv[2];
    if (arg && /^\d{4}-W\d{1,2}$/.test(arg)) {
        const match = arg.match(/^(\d{4})-W(\d{1,2})$/);
        year = parseInt(match[1]);
        week = parseInt(match[2]);
    } else {
        const today = new Date();
        year = today.getFullYear();
        week = getISOWeek(today);
    }
    
    console.log(`📆 ウィークリーノートを生成中: ${year}-W${String(week).padStart(2, '0')}`);
    
    // 出力フォルダを作成
    const outputFolder = path.join(VAULT_PATH, WEEKLY_NOTE_FOLDER);
    if (!fs.existsSync(outputFolder)) {
        fs.mkdirSync(outputFolder, { recursive: true });
        console.log(`📁 フォルダを作成: ${outputFolder}`);
    }
    
    // ウィークリーノートを生成
    const { content, fileName, foundNotes } = generateWeeklyNote(year, week);
    const outputPath = path.join(outputFolder, fileName);
    
    // 既存ファイルのチェック
    if (fs.existsSync(outputPath)) {
        console.log(`⚠️  既存のウィークリーノートが存在します: ${fileName}`);
        console.log(`   上書きを避けるため、スキップします。`);
        return;
    }
    
    // ファイルを書き込み
    fs.writeFileSync(outputPath, content, 'utf8');
    console.log(`✅ ウィークリーノートを作成しました: ${outputPath}`);
    console.log(`   デイリーノート: ${foundNotes}件を集約`);
}

main();

File: 00_Inbox/自動化スクリプト/create_weekly_note.py
#!/usr/bin/env python3
"""
ウィークリーノート自動生成スクリプト
毎週土曜日に実行し、その週（月〜日）のデイリーノートを集約したウィークリーノートを作成

使い方:
  python3 create_weekly_note.py
  python3 create_weekly_note.py 2026-W03  # 特定の週を指定
"""

import os
import sys
import re
from datetime import datetime, timedelta

# ===== 設定 =====
VAULT_PATH = '/Users/sota.yamanaka/Documents/Obsidian Vault'
DAILY_NOTE_FOLDER = '01_Fleeting Note/日記'
WEEKLY_NOTE_FOLDER = '01_Fleeting Note/日記/週次'


def get_iso_week(date):
    """ISO週番号を取得"""
    return date.isocalendar()[1]


def get_iso_year(date):
    """ISO年を取得（週年）"""
    return date.isocalendar()[0]


def get_week_start(year, week):
    """ISO週の開始日（月曜日）を取得"""
    # ISO週1の月曜日を基準に計算
    jan_4 = datetime(year, 1, 4)
    # jan_4が含まれる週の月曜日
    week_1_monday = jan_4 - timedelta(days=jan_4.weekday())
    # 指定週の月曜日
    return week_1_monday + timedelta(weeks=week - 1)


def get_day_of_week(date):
    """日本語曜日を取得"""
    days = ['月', '火', '水', '木', '金', '土', '日']
    return days[date.weekday()]


def parse_daily_note(content):
    """デイリーノートの内容をパース"""
    # YAMLフロントマターを除去
    body_content = re.sub(r'^---[\s\S]*?---\s*', '', content)
    
    sections = {}
    section_order = []
    
    # ##見出しで分割
    parts = re.split(r'^##\s+', body_content, flags=re.MULTILINE)
    
    for part in parts:
        part = part.strip()
        if not part:
            continue
        
        lines = part.split('\n')
        header_name = lines[0].strip()
        body = '\n'.join(lines[1:]).strip()
        
        if not body:
            continue
        
        # 内容が空白行のみ、または "- " だけの行のみの場合はスキップ
        cleaned_body = re.sub(r'^\s*-\s*$', '', body, flags=re.MULTILINE).strip()
        if not cleaned_body:
            continue
        
        if header_name not in sections:
            sections[header_name] = body
            section_order.append(header_name)
    
    return sections, section_order


def generate_weekly_note(year, week):
    """ウィークリーノートを生成"""
    week_start = get_week_start(year, week)
    week_end = week_start + timedelta(days=6)
    
    daily_note_path = os.path.join(VAULT_PATH, DAILY_NOTE_FOLDER)
    
    # セクションごとの集約データ
    accumulated_sections = {}
    all_headers_order = []
    found_notes = 0
    
    # 7日間のデイリーノートを処理
    for i in range(7):
        current_date = week_start + timedelta(days=i)
        date_str = current_date.strftime('%Y-%m-%d')
        day_of_week = get_day_of_week(current_date)
        file_path = os.path.join(daily_note_path, f'{date_str}.md')
        
        if os.path.exists(file_path):
            found_notes += 1
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            sections, section_order = parse_daily_note(content)
            
            for header_name in section_order:
                if header_name not in accumulated_sections:
                    accumulated_sections[header_name] = []
                    all_headers_order.append(header_name)
                
                body = sections[header_name]
                # 引用形式で日付付きで追加
                quoted_body = '\n'.join(f'> {line}' for line in body.split('\n'))
                accumulated_sections[header_name].append(
                    f'> [!quote] [[{date_str}]] ({day_of_week})\n{quoted_body}'
                )
    
    # ノートの生成
    output = f"""---
tags: [weekly-review]
week: {year}-W{week:02d}
date-range: {week_start.strftime('%Y-%m-%d')} ~ {week_end.strftime('%Y-%m-%d')}
created: {datetime.now().strftime('%Y-%m-%d')}
---

# 📆 Weekly Note: {year}-W{week:02d}
**期間:** {week_start.strftime('%Y-%m-%d')} 〜 {week_end.strftime('%Y-%m-%d')}
**デイリーノート数:** {found_notes}/7

---

"""
    
    if not all_headers_order:
        output += '> [!warning] 今週のデイリーノートが見つかりませんでした、または内容が空です。\n'
    else:
        for header in all_headers_order:
            output += f'## {header}\n\n'
            output += '\n\n'.join(accumulated_sections[header])
            output += '\n\n'
    
    # 週の振り返りセクション
    output += """---

## 📊 週の振り返り

### 今週のハイライト
- 

### 来週に向けて
- 

"""
    
    return {
        'content': output,
        'file_name': f'{year}-W{week:02d}.md',
        'found_notes': found_notes
    }


def main():
    """メイン処理"""
    # コマンドライン引数から週を取得、または現在の週を使用
    year = None
    week = None
    
    if len(sys.argv) > 1:
        arg = sys.argv[1]
        match = re.match(r'^(\d{4})-W(\d{1,2})$', arg)
        if match:
            year = int(match.group(1))
            week = int(match.group(2))
    
    if year is None or week is None:
        today = datetime.now()
        year = get_iso_year(today)
        week = get_iso_week(today)
    
    print(f'📆 ウィークリーノートを生成中: {year}-W{week:02d}')
    
    # 出力フォルダを作成
    output_folder = os.path.join(VAULT_PATH, WEEKLY_NOTE_FOLDER)
    if not os.path.exists(output_folder):
        os.makedirs(output_folder, exist_ok=True)
        print(f'📁 フォルダを作成: {output_folder}')
    
    # ウィークリーノートを生成
    result = generate_weekly_note(year, week)
    output_path = os.path.join(output_folder, result['file_name'])
    
    # 既存ファイルのチェック
    if os.path.exists(output_path):
        print(f"⚠️  既存のウィークリーノートが存在します: {result['file_name']}")
        print('   上書きを避けるため、スキップします。')
        return
    
    # ファイルを書き込み
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(result['content'])
    
    print(f'✅ ウィークリーノートを作成しました: {output_path}')
    print(f"   デイリーノート: {result['found_notes']}件を集約")


if __name__ == '__main__':
    main()
